tests/NHSD.BuyingCatalogue.Ordering.Api.IntegrationTests/Steps/EmailSteps.cs

```diff
@@ -37,46 +44,116 @@ public async Task EmailIsNotSent()
-                {
-                    new EmailAddress(Empty, expectedContents.From)
-                },
+                From = new List<EmailAddress> { new EmailAddress(Empty, expectedContents.From) },
+                To = new List<EmailAddress> { new EmailAddress(Empty, expectedContents.To) },
                 expectedContents.Text,
                 expectedContents.Subject,
-                To = new List<EmailAddress>
-                {
-                    new EmailAddress(Empty, expectedContents.To)
-                }
             };
-            
+
             var actual = (await _emailServerDriver.FindAllEmailsAsync()).First();
 
             actual.Should().BeEquivalentTo(expected, conf => conf.IncludingAllDeclaredProperties());
         }
 
+        [Then(@"the patient numbers price type attachment contains the correct information")]
+        public async Task ThenThePatientNumbersPriceTypeAttachmentContainsTheCorrectInformation()
+        {
+            string[] patientNumbersCsvHeader = {
+                "Call Off Agreement ID",
+                "Call Off Ordering Party ID",
+                "Call Off Ordering Party Name",
+                "Call Off Commencement Date",
+                "Service Recipient ID",
+                "Service Recipient Name",
+                "Service Recipient Item ID",
+                "Supplier ID",
+                "Supplier Name",
+                "Product ID",
+                "Product Name",
+                "Product Type",
+                "Quantity Ordered",
+                "Unit of Order",
+                "Price",
+                "Funding Type",
+                "M1 planned (Delivery Date)",
+                "Actual M1 date",
+                "Buyer verification date (M2)",
+                "Cease Date"
+            };
+
+            var actual = (await _emailServerDriver.FindAllEmailsAsync()).First().Attachments.First();
+
+            var attachmentData = new MemoryStream(actual.AttachmentData.ToArray());
+            var csvText = new TextFieldParser(attachmentData);
+            csvText.SetDelimiters(",");
+
+            csvText.ReadFields().Should().BeEquivalentTo(patientNumbersCsvHeader);
+
+            var orderId = _scenarioContext[ScenarioContextKeys.OrderId].ToString();
+
+            var order = _orderContext.OrderReferenceList.GetByOrderId(orderId);
+            var orderItems = _orderContext.OrderItemReferenceList.FindByOrderId(orderId).ToArray();
+            var serviceRecipients = _orderContext.ServiceRecipientReferenceList.FindByOrderId(orderId).ToList();
+
+            foreach (var orderItem in orderItems)
+            {
+                csvText.EndOfData.Should().BeFalse();
+
+                var row = csvText.ReadFields();
+
+                var data = CreatePatientNumbersTypeAttachmentData(order, orderItem, serviceRecipients);
+
+                row.Should().BeEquivalentTo(data);
+            }
+
+            csvText.EndOfData.Should().BeTrue();
```

Comment:
Rather than asserting each row of the email attachment, do you think it would be better to include one assert statement that checks the two complete lists? Meaning generate a list of string[] for the actual data and expected data. This would hopefully make the test easier to read & if there was more than one errors in the document the test failure result would contain all issues not use the first row to contain a mismatch.

Reply:
I guess so. The reason for implementing it like this, was being more explicit about what the error is. E.g. if the file still had another row in, then the endOfData would be false

Reply:
*I still prefer this method tho